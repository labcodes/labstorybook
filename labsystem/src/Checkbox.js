import React from "react";
import PropTypes, { string, number, bool } from "prop-types";
import { isUndefined } from "lodash";

import Icon from "./Icon";

// Checkbox //

export default class Checkbox extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    checked: PropTypes.bool,
    value: PropTypes.oneOfType([string, number, bool]),
    indeterminate: PropTypes.bool,
    defaultChecked: PropTypes.bool,
    className: PropTypes.string,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    disabled: false,
    checked: undefined,
    value: null,
    indeterminate: false,
    defaultChecked: undefined,
    className: undefined,
    onChange: () => {},
  };

  constructor(props) {
    super(props);
    const { defaultChecked, checked, id } = props;
    if (!isUndefined(defaultChecked) && !isUndefined(checked)) {
      // eslint-disable-next-line no-console
      console.warn(
        `You are setting both checked and defaultChecked for input ${id} at the same time. We always initialize the checkbox with defaultChecked. Make sure this is the behaviour you want.`
      );
    }

    let localChecked = false;
    if (defaultChecked) {
      localChecked = defaultChecked;
    } else if (!isUndefined(checked)) {
      localChecked = checked;
    }

    this.state = {
      localChecked,
    };
  }

  componentDidUpdate(prevProps) {
    const { checked } = this.props;
    if (checked !== prevProps.checked) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState(() => ({ localChecked: checked }));
    }
  }

  checkIcon = () => {
    const { disabled, indeterminate } = this.props;
    let color = "white";
    let type = "check";

    if (indeterminate) {
      type = "minus";
    }
    if (disabled) {
      color = "mineral-40";
    }

    return <Icon type={type} color={color} size="small" />;
  };

  handleOnChange = (event) => {
    const { onChange } = this.props;
    if (!isUndefined(onChange)) {
      onChange(event);
    }

    this.setState((state) => ({ localChecked: !state.localChecked }));
  };

  render() {
    const {
      id,
      name,
      className,
      label,
      disabled,
      indeterminate,
      value,
    } = this.props;

    const { localChecked } = this.state;

    return (
      <React.Fragment>
        <input
          className={`lab-checkbox ${className || ""}`}
          type="checkbox"
          id={id}
          name={name}
          checked={localChecked}
          onChange={this.handleOnChange}
          {...(value ? { value } : undefined)}
          {...(disabled ? { disabled } : undefined)}
        />
        <label className="lab-checkbox__label" htmlFor={id}>
          <span className="lab-checkbox__box">
            {localChecked || indeterminate ? this.checkIcon() : ""}
          </span>
          {label}
        </label>
      </React.Fragment>
    );
  }
}
