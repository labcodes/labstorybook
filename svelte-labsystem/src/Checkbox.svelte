<script>
  import Icon from "./Icon.svelte";

  export let id;
  export let name;
  export let label;
  export let checked = undefined;
  export let value = undefined;
  export let disabled = undefined;
  export let indeterminate = false;

  let className = "";
  export { className as class };
</script>

<input
  class={`lab-checkbox${` ${className}` || ''}`}
  type="checkbox"
  {id}
  {name}
  {value}
  {disabled}
  bind:checked
  on:change />

<label class="lab-checkbox__label" for={id}>
  <span class="lab-checkbox__box">
    {#if indeterminate && checked}
      <Icon
        type="minus"
        size="small"
        color={disabled ? 'mineral-40' : 'white'} />
    {:else if checked}
      <Icon
        type="check"
        size="small"
        color={disabled ? 'mineral-40' : 'white'} />
    {/if}
  </span>

  {label}
</label>
