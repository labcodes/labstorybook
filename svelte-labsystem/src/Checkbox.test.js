import { render } from "@testing-library/svelte";
import Checkbox from "./Checkbox.svelte";

describe("Checkbox", () => {
  it("renders with base props", async () => {
    const { container } = render(Checkbox, {
      name: "test",
      id: "test",
      label: "Test",
    });
    expect(container).toMatchSnapshot();
  });

  it("renders as expected when passing disabled as true", async () => {
    const { container } = render(Checkbox, {
      name: "test",
      id: "test",
      label: "Test",
      disabled: true,
    });

    expect(container).toMatchSnapshot();
    expect(container.querySelector("input").disabled).toBeTruthy();
  });

  it("renders as expected when passing indeterminate as true", async () => {
    const { container } = render(Checkbox, {
      name: "test",
      id: "test",
      label: "Test",
      indeterminate: true,
    });

    expect(container).toMatchSnapshot();
    expect(container.querySelector("input").indeterminate).toBeTruthy();
  });

  it("renders as expected when passing a different class", async () => {
    const { container } = render(Checkbox, {
      name: "test",
      id: "test",
      label: "Test",
      class: "dummy-class",
    });
    expect(container).toMatchSnapshot();
  });
});
