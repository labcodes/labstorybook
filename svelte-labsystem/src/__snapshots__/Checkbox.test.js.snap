// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders as expected when passing a different class 1`] = `
<body>
  <div>
    <input
      class="lab-checkbox dummy-class"
      id="test"
      name="test"
      type="checkbox"
      value="undefined"
    />
     
    <label
      class="lab-checkbox__label"
      for="test"
    >
      <span
        class="lab-checkbox__box"
      />
       
      Test
    </label>
  </div>
</body>
`;

exports[`Checkbox renders as expected when passing disabled as true 1`] = `
<body>
  <div>
    <input
      class="lab-checkbox "
      disabled=""
      id="test"
      name="test"
      type="checkbox"
      value="undefined"
    />
     
    <label
      class="lab-checkbox__label"
      for="test"
    >
      <span
        class="lab-checkbox__box"
      />
       
      Test
    </label>
  </div>
</body>
`;

exports[`Checkbox renders as expected when passing indeterminate as true 1`] = `
<body>
  <div>
    <input
      class="lab-checkbox "
      id="test"
      name="test"
      type="checkbox"
      value="undefined"
    />
     
    <label
      class="lab-checkbox__label"
      for="test"
    >
      <span
        class="lab-checkbox__box"
      />
       
      Test
    </label>
  </div>
</body>
`;

exports[`Checkbox renders with base props 1`] = `
<body>
  <div>
    <input
      class="lab-checkbox "
      id="test"
      name="test"
      type="checkbox"
      value="undefined"
    />
     
    <label
      class="lab-checkbox__label"
      for="test"
    >
      <span
        class="lab-checkbox__box"
      />
       
      Test
    </label>
  </div>
</body>
`;
